fluxcat/core.py

```diff
@@ -0,0 +1,1510 @@
+            self.param = param.tolist()
+            self.param_cov = param_cov.tolist()
+
+            # Calculate statistics
+            if not isinstance(self.stats, dict):
+                self.stats = {}
+            self.stats["ndof"] = len(x) - self.nparam
+            self.stats["chisq"] = np.sum(resid**2 / np.diag(C))
+
+        # Return results
+        return self.param, self.param_cov, self.stats
+
+    def _get_x(self, freq):
+        return np.log(freq / self.freq_pivot)
+
+    @staticmethod
+    def _vandermonde(x, nparam):
+        return np.vstack(tuple([x**rank for rank in range(nparam)])).T
+
+    @staticmethod
+    def _fit_func(x, *param):
+        return param[0] * np.exp(
+            np.sum(
+                [par * x ** (rank + 1) for rank, par in enumerate(param[1:])], axis=0
+            )
+        )
+
+    @staticmethod
+    def _deriv_fit_func(x, *param):
+        z = param[0] * np.exp(
+            np.sum(
+                [par * x ** (rank + 1) for rank, par in enumerate(param[1:])], axis=0
+            )
+        )
+
+        dfdp = np.array([z * x**rank for rank in range(len(param))])
+        dfdp[0] /= param[0]
+
+        return dfdp
+
+
+class MetaFluxCatalog(type):
+    """Metaclass for FluxCatalog.  Defines magic methods
+    for the class that can act on and provide access to the
+    catalog of all astronomical sources.
+    """
+
+    def __str__(self):
+        return self.string()
+
+    def __iter__(self):
+        return self.iter()
+
+    def __reversed__(self):
+        return self.reversed()
+
+    def __len__(self):
+        return self.len()
+
+    def __getitem__(self, key):
+        return self.get(key)
+
+    def __contains__(self, item):
+        try:
+            obj = self.get(item)
+        except KeyError:
+            obj = None
+
+        return obj is not None
+
+
+class FluxCatalog(metaclass=MetaFluxCatalog):
+    """Class for cataloging astronomical sources and predicting
+    their flux density at radio frequencies based on spectral fits
+    to previous measurements.
+
+    Class methods act upon and provide access to the catalog of
+    all sources.  Instance methods act upon and provide access
+    to individual sources.  All instances are stored in an
+    internal class dictionary.
+
+    Attributes
+    ----------
+    fields : list
+        List of attributes that are read-from and written-to the
+        JSON catalog files.
+    model_lookup : dict
+        Dictionary that provides access to the various models that
+        can be fit to the spectrum.  These models should be
+        subclasses of FitSpectrum.
+    """
+
+    fields = [
+        "ra",
+        "dec",
+        "alternate_names",
+        "model",
+        "model_kwargs",
+        "stats",
+        "param",
+        "param_cov",
+        "measurements",
+    ]
+
+    model_lookup = {"CurvedPowerLaw": CurvedPowerLaw}
+
+    _entries = {}
+    _collections = {}
+    _alternate_name_lookup = {}
```

Comment:
Thinking about HFBcat, if we want to make this its own package, we really need to fix the way FluxCat is storing its catalogues (see the problem I addressed here for HFBCat: https://github.com/chime-experiment/ch_util/pull/86 ).  I think the easy solution is to move these class variables its using to store the catalogues it loads at import time out of the class and turn them into module-level attributes.

Reply:
I think that's a reasonable idea. Regardless, I don't think it's a great idea to be loading catalogs at import time. Maybe `FluxCatalog` could load from `DEFAULT_COLLECTIONS` when it's initialized instead?

Reply:
Actually, I think that's fixed here. Now that I think about it, I need to check if there's a reliance on this load-at-import elsewhere.